warehouse/query-core/src/main/java/datawave/query/jexl/visitors/BooleanOptimizationRebuildingVisitor.java

```diff
@@ -54,27 +50,22 @@ protected JexlNode optimizeTree(JexlNode currentNode, JexlNode newNode, Object d
             Tuple2<JexlNode,JexlNode> nodes = prune(currentNode, andNode, orNode);
             
             JexlNode prunedNode = nodes.first();
-            
             JexlNode toAttach = nodes.second();
             toAttach = TreeFlatteningRebuildingVisitor.flatten(toAttach);
+            ASTOrNode newNode = new ASTOrNode(ParserTreeConstants.JJTORNODE);
+            newNode.image = currentNode.image;
             
             for (int i = 0; i < toAttach.jjtGetNumChildren(); i++) {
                 JexlNode node = copy(prunedNode);
                 JexlNode attach = (JexlNode) toAttach.jjtGetChild(i).jjtAccept(this, data);
+                attach.jjtSetParent(node);
```

Comment:
I see lots of refactoring here but this is the only line that actually fixes a parenting problem for this class.  Refactoring for the sake of refactoring may do more harm than good if anybody else is working on these visitors for any other reason.  It also makes it very difficult for anybody to actually accept the entire package and merge.  I think you may need to pull back a little from the refactoring efforts.  Better yet a separate pull request per visitor that is only refactoring would be considerably more palatable.